apps/extension/src/storage/base.ts

```diff
@@ -20,30 +20,55 @@ export interface StorageItem<T> {
 }
 
 type Version = string;
+
+export type MigrationMap<OldState, NewState> = {
+  [K in keyof OldState & keyof NewState]?: (
+    prev: OldState[K],
+    get: <K extends keyof NewState>(key: K) => Promise<NewState[K]>,
+  ) => NewState[K] | Promise<NewState[K]>;
+};
+
 // It is quite difficult writing a generic that covers all migration function kinds.
 // Therefore, the writer of the migration should ensure it is typesafe when they define it.
-// See `migration.test.ts` for an example.
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-type Migration = Record<Version, (a: any) => any>;
-type Migrations<K extends string | number | symbol> = Partial<Record<K, Migration>>;
+export type Migrations<T> = Partial<Record<Version, MigrationMap<any, T>>>;
+
+export type VersionSteps = Record<Version, Version>;
 
 export class ExtensionStorage<T> {
```

Comment:
this is not a suggested code change, but rather a general statement if we decide to restructure our migration mechanism in the future. 

personally, migrations feel more transactional in nature, where migrations for multiple fields should ideally be handled atomically, rather than the current approach of performing migrations on individual fields separately.

one potential approach is a _batch_ migration mechanism, where you define migrations at the version level. It ensures atomicity by checking that all changes associated with a particular version are applied together in a single transaction. so either all migrations succeed or none do

Reply:
This is an important comment. It's unfortunate that chrome.local.storage does not come with any support for versioning or migrations. For this reason, we had to create our own custom migration code. It was probably a bad design that each field has their own version key. A possibly better design is to save a global version number, check for needed migrations when someone attempts a `get()`, and if so, migrate all keys at once.

Going to spend today trying this out. The hardest part of this will be migrating from the old data structure to the new.